File: src/main/java/de/klierlinge/clear19/widgets/geometry/Point.java
package de.klierlinge.clear19.widgets.geometry;

public class Point
{
    public static final Point ZERO = new Point(0, 0);
    
    private final int x;
    private final int y;
    
    public Point(int x, int y)
    {
        super();
        this.x = x;
        this.y = y;
    }
    
    final public int getX()
    {
        return x;
    }
    
    final public int getY()
    {
        return y;
    }

    public Point add(Point other)
    {
        return new Point(getX() + other.getX(), getY() + other.getY());
    }
    
    public Point reversed()
    {
        return new Point(-x, -y);
    }
    
    public AnchoredPoint anchored(Anchor anchor)
    {
        return new AnchoredPoint(getX(), getY(), anchor);
    }

    @Override
    public int hashCode()
    {
        final int prime = 31;
        int result = 1;
        result = prime * result + x;
        result = prime * result + y;
        return result;
    }

    @Override
    public boolean equals(Object obj)
    {
        if(this == obj)
            return true;
        if(obj == null)
            return false;
        if(getClass() != obj.getClass())
            return false;
        Point other = (Point)obj;
        if(x != other.x)
            return false;
        if(y != other.y)
            return false;
        return true;
    }

    public Size toSize()
    {
        return new Size(getX(), getY());
    }
    
    @Override
    public String toString()
    {
        return "Point [x=" + x + ", y=" + y + "]";
    }
}

File: clear19/widgets/geometry/point.py
from __future__ import annotations

from builtins import abs
from dataclasses import dataclass
from typing import Iterator

from clear19.widgets.geometry.anchor import Anchor
from clear19.widgets.geometry.size import Size


@dataclass
class Point:
    x: float
    y: float

    def anchored(self, anchor: Anchor) -> AnchoredPoint:
        return AnchoredPoint(self.x, self.y, anchor)

    def __sub__(self, other: Point) -> Size:
        return Size(abs(self.x - other.x), abs(self.y - other.y))

    def __iter__(self) -> Iterator[float]:
        return iter((self.x, self.y))


class AnchoredPoint(Point):
    _anchor: Anchor

    def __init__(self, x: float, y: float, anchor: Anchor):
        super().__init__(x, y)
        self._anchor = anchor

    @property
    def anchor(self) -> Anchor:
        return self._anchor


ZERO: Point = Point(0, 0)
ZERO_TOP_LEFT: AnchoredPoint = AnchoredPoint(0, 0, Anchor.TOP_LEFT)
